src/java.base/aix/classes/sun/nio/ch/AixPollPort.java

```diff
@@ -303,11 +307,23 @@ private class EventHandlerTask implements Runnable {
         private Event poll() throws IOException {
             try {
                 for (;;) {
-                    int n;
+                    int n, m;
                     controlLock.lock();
                     try {
-                        n = Pollset.pollsetPoll(pollset, address,
+                        m = n = Pollset.pollsetPoll(pollset, address,
                                      MAX_EVENTS_TO_POLL, Pollset.PS_NO_TIMEOUT);
+                        while (m-- > 0) {
+                            long eventAddress = Pollset.getEvent(address, m);
+                            int fd = Pollset.getDescriptor(eventAddress);
+
+                            // To emulate one shot semantic we need to remove
+                            // the file descriptor here.
+                            if (fd != sp[0] && fd != ctlSp[0]) {
+                                synchronized (controlQueue) {
```

Comment:
Ok, the lock order `controlLock` before `controlQueue` seems to be consistent. Otherwise, we might get deadlocks. (Exception is `controlLock.tryLock()` which is ok.)
Shouldn't `synchronized (controlQueue)` better get removed from `processControlQueue()`? The lock has always been acquired before it is called. If it didn't, it would make the lock order inconsistent and possibly cause deadlocks.

Reply:
Removed the additional` synchronized (controlQueue)` from `processControlQueue()`